android/library/src/main/java/io/logto/android/Logto.kt

```diff
@@ -41,71 +44,98 @@ class Logto(
+            return
         }
-    } ?: throw LogtoException(LogtoException.NOT_AUTHENTICATED)
 
-    fun getIdTokenClaims(): JwtClaims {
-        return tokenSet?.getIdTokenClaims()
-            ?: throw LogtoException(LogtoException.NOT_AUTHENTICATED)
+        refreshTokenSet { exception, newTokenSet ->
+            block(exception, newTokenSet?.accessToken)
+        }
+    }
+
+    fun getIdTokenClaims(block: HandleIdTokenClaimsCallback) {
+        if (tokenSet == null) {
+            block(LogtoException(LogtoException.NOT_AUTHENTICATED), null)
+            return
+        }
+        try {
+            val idTokenClaims = tokenSet!!.getIdTokenClaims()
+            block(null, idTokenClaims)
+        } catch (exception: LogtoException) {
+            block(exception, null)
+        }
     }
 
     fun signInWithBrowser(
         context: Context,
-        onComplete: (tokenSet: TokenSet) -> Unit,
-    ) = AuthManager.start(
-        context,
-        BrowserSignInFlow(
+        onComplete: HandleTokenSetCallback,
+    ) {
+        val signInFlow = BrowserSignInFlow(
             logtoAndroidClient,
         ) { exception, tokenSet ->
             if (exception != null) {
-                throw exception
+                onComplete(exception, null)
+                return@BrowserSignInFlow
             }
             if (tokenSet == null) {
-                throw LogtoException(LogtoException.UNKNOWN_ERROR)
+                onComplete(LogtoException(LogtoException.UNKNOWN_ERROR), null)
+                return@BrowserSignInFlow
             }
             updateTokenSet(tokenSet)
             AuthManager.reset()
-            onComplete(tokenSet)
+            onComplete(null, tokenSet)
         }
-    )
+
+        AuthManager.start(context, signInFlow)
+    }
 
     fun signOutWithBrowser(
         context: Context,
-    ) = tokenSet?.let {
-        AuthManager.start(
-            context,
-            BrowserSignOutFlow(
-                idToken = it.idToken,
-                logtoAndroidClient = logtoAndroidClient,
-            ) { exception ->
-                updateTokenSet(null)
-                AuthManager.reset()
-                if (exception != null) {
-                    throw exception
-                }
-            }
-        )
-    } ?: throw LogtoException(LogtoException.NOT_AUTHENTICATED)
+        block: HandleLogtoExceptionCallback? = null
+    ) {
+        if (tokenSet == null) {
+            block?.invoke(LogtoException(LogtoException.NOT_AUTHENTICATED))
+            return
+        }
 
-    fun refreshTokenSet(
-        block: (tokenSet: TokenSet) -> Unit,
-    ) = tokenSet?.let {
-        val refreshToken = it.refreshToken
-            ?: throw LogtoException(LogtoException.REFRESH_TOKEN_IS_NOT_SUPPORTED)
+        val browserSignOutFlow = BrowserSignOutFlow(
+            idToken = tokenSet!!.idToken,
+            logtoAndroidClient = logtoAndroidClient,
+        ) { exception ->
+            updateTokenSet(null)
```

Comment:
do we need to check the exception first before clear the tokenSet?

Reply:
post discussion, we clear local tokenSet once user chose to logout